compiler/destroyer.nim

```diff
@@ -399,10 +411,14 @@ proc pArg(arg: PNode; c: var Con; isSink: bool): PNode =
     elif arg.kind in {nkBracket, nkObjConstr, nkTupleConstr, nkBracket, nkCharLit..nkFloat128Lit}:
       discard "object construction to sink parameter: nothing to do"
       result = arg
-    elif arg.kind == nkSym and (isSinkParam(arg.sym) or
-         arg.sym.kind in InterestingSyms and isLastRead(arg, c)):
-      # it is the last read be final consumption. We need to reset the memory
-      # to disable the destructor which we have not elided:
+    elif arg.kind == nkSym and isSinkParam(arg.sym):
+      # Sinked params can be consumed only once. We need to reset the memory
+      # to disable the destructor which we have not elided
+      result = destructiveMoveVar(arg, c)
+      sinkParamConsumed(c, arg)
```

Comment:
Why not use ``isLastRead(arg, c)`` here instead of your control-flow-ignoring ``sinkParamConsumed(c, arg)``?

Reply:
Good, idea. With minor modifications to the dfa I managed to make it working